File: backend/storage.py
import os
import json
import base64
import mimetypes
import shutil
import time
import re
import secrets
import string
from flask import request, Response
from crypto import process_chunk, get_stream_decryptor
from config import VAULT_DIR, VAULT_IMAGES, VAULT_VIDEOS, VAULT_FILES, THUMBNAIL_DIR, TMP_DIR

def save_meta(folder: str, meta: dict):
    with open(os.path.join(folder, "meta.json"), "w") as f:
        json.dump(meta, f)

def get_meta(folder: str) -> dict:
    with open(os.path.join(folder, "meta.json"), "r") as f:
        return json.load(f)

def is_vault_item(path: str) -> bool:
    return os.path.isdir(path) and os.path.exists(os.path.join(path, "meta.json"))

def vault_data_path(folder: str) -> str:
    # Look for monolithic file first, then legacy c0000.dat
    monolith = os.path.join(folder, "data.enc")
    if os.path.exists(monolith): return monolith
    legacy = os.path.join(folder, "c0000.dat")
    if os.path.exists(legacy): return legacy
    return monolith # default to monolith for new writes

def find_item_path(enc_name: str) -> str:
    """Helper to locate an encrypted folder in any vault subfolder."""
    for sub in ["images", "videos", "documents", "files"]:
        p = os.path.join(VAULT_DIR, sub, enc_name)
        if is_vault_item(p): return p
    # Fallback to root vault (for old files before restructure)
    p_root = os.path.join(VAULT_DIR, enc_name)
    if is_vault_item(p_root): return p_root
    return None

def write_encrypted_to_file(data: bytes, nonce: bytes, global_offset: int, folder: str):
    """Appends/Writes encrypted data to the single monolithic file."""
    encrypted = process_chunk(data, nonce, global_offset)
    dp = vault_data_path(folder)
    mode = "r+b" if os.path.exists(dp) else "wb"
    with open(dp, mode) as f:
        f.seek(global_offset)
        f.write(encrypted)

def finalize_vault_item(temp_dir, original_name, nonce, total_size):
    """Moves temp upload to the correct vault subfolder with metadata."""
    enc_name = os.path.basename(temp_dir)
    mime = mimetypes.guess_type(original_name)[0] or "application/octet-stream"
    
    # ── RANDOMIZE FILENAME ──────────────────────────────────────────────────
    ext = os.path.splitext(original_name)[1]
    rand_base = ''.join(secrets.choice(string.ascii_lowercase + string.digits) for _ in range(12))
    random_name = f"vault_{rand_base}{ext}"
    
    # Determine correct subfolder
    sub = "files"
    if mime.startswith("image/"): sub = "images"
    elif mime.startswith("video/"): sub = "videos"
    elif any(d in mime for d in ["pdf", "text", "msword", "document", "spreadsheet", "presentation", "csv"]):
        sub = "documents"
    
    final_dir = os.path.join(VAULT_DIR, sub, enc_name)
    os.makedirs(os.path.dirname(final_dir), exist_ok=True)
    
    meta = {
        "name":        random_name,
        "original":    original_name,
        "size":        total_size,
        "nonce":       base64.b64encode(nonce).decode(),
        "type":        mime,
        "created_at":  time.time()
    }
    save_meta(temp_dir, meta)
    
    state_path = os.path.join(temp_dir, ".state")
    if os.path.exists(state_path): os.remove(state_path)
    
    if os.path.exists(final_dir): shutil.rmtree(final_dir)
    os.rename(temp_dir, final_dir)
    
    # Generate thumbnail
    if mime.startswith("image/") or mime.startswith("video/"):
        generate_thumbnail(enc_name)

def generate_thumbnail(enc_name):
    """Generates a mid-quality, encrypted thumbnail (image or video). 600px/75%."""
    from PIL import Image
    import io
    
    vault_item_path = find_item_path(enc_name)
    if not vault_item_path: return
    
    meta = get_meta(vault_item_path)
    nonce = base64.b64decode(meta["nonce"])
    dp = vault_data_path(vault_item_path)
    if not os.path.exists(dp): return
    
    mime = meta.get("type", "")
    thumb_bytes = None
    
    if mime.startswith("image/"):
        thumb_bytes = _extract_image_thumb(dp, nonce)
    elif mime.startswith("video/"):
        thumb_bytes = _extract_video_thumb(dp, nonce, enc_name)
        
    if thumb_bytes:
        # Store encrypted thumbnail
        th_path = os.path.join(THUMBNAIL_DIR, enc_name)
        th_nonce = secrets.token_bytes(16)
        with open(th_path, "wb") as tf:
            tf.write(th_nonce + process_chunk(thumb_bytes, th_nonce, 0))

def _extract_image_thumb(dp, nonce):
    from PIL import Image
    import io
    # Decrypt first 50MB
    with open(dp, "rb") as f:
        enc_data = f.read(50 * 1024 * 1024)
        dec = get_stream_decryptor(nonce, 0)
        raw_data = dec.update(enc_data)
        
    try:
        img = Image.open(io.BytesIO(raw_data))
        img.thumbnail((600, 600))
        if img.mode in ("RGBA", "P"):
            img = img.convert("RGB")
        out = io.BytesIO()
        img.save(out, format="JPEG", quality=75)
        return out.getvalue()
    except Exception as e:
        print(f"[Vault] Image thumb failed: {e}")
        return None

def _extract_video_thumb(dp, nonce, enc_name):
    import cv2
    import numpy as np
    from PIL import Image
    import io
    
    # Decrypt first 20MB to a temporary file for OpenCV to read
    temp_vid = os.path.join(TMP_DIR, f"th_tmp_{enc_name}")
    try:
        with open(dp, "rb") as f:
            enc_header = f.read(20 * 1024 * 1024)
            dec = get_stream_decryptor(nonce, 0)
            raw_header = dec.update(enc_header)
            with open(temp_vid, "wb") as tv:
                tv.write(raw_header)
        
        cap = cv2.VideoCapture(temp_vid)
        # Try to seek to skip potential black frames at start (1 second)
        cap.set(cv2.CAP_PROP_POS_MSEC, 1000)
        success, frame = cap.read()
        if not success:
            # Fallback to first frame
            cap.set(cv2.CAP_PROP_POS_MSEC, 0)
            success, frame = cap.read()
            
        cap.release()
        if os.path.exists(temp_vid): os.remove(temp_vid)
        
        if success:
            # Convert BGR (CV2) to RGB (PIL)
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            img = Image.fromarray(frame_rgb)
            img.thumbnail((600, 600))
            out = io.BytesIO()
            img.save(out, format="JPEG", quality=75)
            return out.getvalue()
            
    except Exception as e:
        if os.path.exists(temp_vid): os.remove(temp_vid)
        print(f"[Vault] Video thumb failed: {e}")
    return None

def serve_vault_stream(enc_name: str, is_download=False):
    """Stream monolithic vault item with HTTP Range support."""
    folder = find_item_path(enc_name)
    if not folder: return "Not found", 404
    
    meta          = get_meta(folder)
    dp            = vault_data_path(folder)
    if not os.path.exists(dp): return "No data", 404

    total_size    = meta["size"]
    original_name = meta["name"]
    nonce         = base64.b64decode(meta["nonce"])
    
    range_header = request.headers.get("Range")
    start, end   = 0, total_size - 1
    status_code  = 200

    if range_header:
        m = re.search(r"bytes=(\d+)-(\d*)", range_header)
        if m:
            start = int(m.group(1))
            # Only cap range if it's a video (for performance). For images/others, 
            # serve the full remaining file if end is not provided.
            if "video" in mime:
                default_end = min(start + 32*1024*1024 - 1, total_size - 1)
            else:
                default_end = total_size - 1
                
            end = int(m.group(2)) if m.group(2) else default_end
            end = min(end, total_size - 1)
        status_code = 206

    length = end - start + 1

    def generate():
        remaining = length
        print(f"[Stream] Starting {mime} ({original_name}), Range: {start}-{end}, Length: {length}")
        try:
            with open(dp, "rb") as f:
                f.seek(start)
                dec = get_stream_decryptor(nonce, start)
                while remaining > 0:
                    chunk_to_read = min(remaining, 4*1024*1024)
                    data = f.read(chunk_to_read)
                    if not data: 
                        print(f"[Stream] Unexpected EOF! Needed {remaining} more bytes.")
                        break
                    
                    out = dec.update(data)
                    yield out
                    remaining -= len(data)
                
                # Force clean-up
                try: dec.finalize()
                except: pass
                print(f"[Stream] Finished {original_name} successfully.")
        except Exception as e:
            print(f"[Stream] Failed during {original_name}: {e}")

    mime = meta["type"]
    headers = {
        "Content-Length": str(length),
        "Accept-Ranges":  "bytes",
        "Content-Type":   mime,
        "Cache-Control":  "no-cache",
    }
    if is_download:
        safe = original_name.replace('"', "_")
        headers["Content-Disposition"] = f'attachment; filename="{safe}"'
    if status_code == 206:
        headers["Content-Range"] = f"bytes {start}-{end}/{total_size}"
    
    return Response(generate(), status=status_code, headers=headers)

File: backend/src/storage.ts
import fs from 'fs-extra';
import path from 'path';
import crypto from 'crypto';
import mime from 'mime-types';
import { pipeline } from 'stream/promises';
import { Transform, PassThrough } from 'stream';

import { VAULT_DIR, THUMBNAIL_DIR, TMP_DIR, ACCESS_KEY } from './config.js';
import {
    getAesKey,
    getCipherAtOffset,
    getDecipherAtOffset,
    processChunk
} from './crypto.js';

import {
    generateThumbnail,
    normalizeVideo,
    getVideoDuration
} from './media.js';

const KEY_BUFFER = getAesKey(ACCESS_KEY);

export interface FileMeta {
    id: string;
    name: string;
    original: string;
    size: number;
    nonce: string;
    type: string;
    created_at: number;
    duration?: number;
    status?: "ready" | "processing" | "error";
    encLevel?: 0 | 1 | 2;
    isEncrypted?: boolean;
    thumb?: boolean;
    subtitles?: { index: number, label: string, lang: string }[];
    userKeyId?: string; // Fingerprint of the personal key (for isolation)
}

export async function saveMeta(folder: string, meta: FileMeta) {
    await fs.writeJson(path.join(folder, "meta.json"), meta);
}

export async function getMeta(folder: string): Promise<FileMeta> {
    return await fs.readJson(path.join(folder, "meta.json"));
}

export async function isVaultItem(p: string): Promise<boolean> {
    return (await fs.pathExists(p)) &&
           (await fs.stat(p)).isDirectory() &&
           (await fs.pathExists(path.join(p, "meta.json")));
}

export function vaultDataPath(folder: string): string {
    const monolith = path.join(folder, "data.enc");
    if (fs.existsSync(monolith)) return monolith;
    const legacy = path.join(folder, "c0000.dat");
    if (fs.existsSync(legacy)) return legacy;
    return monolith;
}

export async function findItemPath(enc_name: string): Promise<string | null> {
    const subfolders = ["images", "videos", "music", "documents", "files"];
    for (const sub of subfolders) {
        const p = path.join(VAULT_DIR, sub, enc_name);
        if (await isVaultItem(p)) return p;
    }
    const p_root = path.join(VAULT_DIR, enc_name);
    if (await isVaultItem(p_root)) return p_root;
    return null;
}

export async function getDetailedListing(v_type: string = "all") {
    const subfolders = ["images", "videos", "music", "documents", "files"];
    let detailed: FileMeta[] = [];
    for (const sub of subfolders) {
        const sub_path = path.join(VAULT_DIR, sub);
        if (!(await fs.pathExists(sub_path))) continue;
        const entries = await fs.readdir(sub_path);
        for (const n of entries) {
            const p = path.join(sub_path, n);
            if (await isVaultItem(p)) {
                try {
                    const m = await getMeta(p);
                    m.id = n;
                    detailed.push(m);
                } catch { continue; }
            }
        }
    }
    if (VAULT_DIR && await fs.pathExists(VAULT_DIR)) {
        const rootEntries = await fs.readdir(VAULT_DIR);
        for (const n of rootEntries) {
            if (subfolders.includes(n)) continue;
            const p = path.join(VAULT_DIR, n);
            if (await isVaultItem(p)) {
                try {
                    const m = await getMeta(p);
                    m.id = n;
                    detailed.push(m);
                } catch { continue; }
            }
        }
    }
    detailed.sort((a, b) => b.created_at - a.created_at);
    if (v_type === "images") detailed = detailed.filter(d => d.type.includes("image"));
    else if (v_type === "videos") detailed = detailed.filter(d => d.type.includes("video"));
    else if (v_type === "music") detailed = detailed.filter(d => d.type.includes("audio"));
    return detailed;
}

export async function writeToVault(
    data: Buffer,
    key: Buffer | null,
    nonce: Buffer,
    globalOffset: number,
    folder: string
) {
    const dp = vaultDataPath(folder);
    const finalData = key ? processChunk(data, key, nonce, globalOffset) : data;
    await fs.ensureFile(dp);
    const flags = globalOffset === 0 ? 'w' : 'r+';
    const fd = await fs.open(dp, flags).catch(() => fs.open(dp, 'r+'));
    try {
        await fs.write(fd, finalData, 0, finalData.length, globalOffset);
    } finally {
        await fs.close(fd);
    }
}

export async function indexMoofOffsets(dp: string, nonce: Buffer, decryptKey: Buffer | null = null): Promise<number[]> {
    const offsets: number[] = [];
    const decipher = decryptKey ? getDecipherAtOffset(decryptKey, nonce, 0) : null;
    const MOOF = Buffer.from('moof');
    let position = 0;
    let buffer = Buffer.alloc(0);

    const scanner = new Transform({
        transform(chunk, _, cb) {
            const bufferChunk = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk as any);
            const decrypted = decipher ? decipher.update(bufferChunk) : bufferChunk;
            buffer = Buffer.concat([buffer, decrypted]);
            let idx;
            while ((idx = buffer.indexOf(MOOF)) !== -1) {
                const start = position + idx - 4;
                if (start >= 0) offsets.push(start);
                buffer = buffer.slice(idx + 4);
                position += idx + 4;
            }
            if (buffer.length > 8) {
                const trim = buffer.length - 8;
                buffer = buffer.slice(-8);
                position += trim;
            }
            cb();
        },
        flush(cb) {
            if (decipher) decipher.final();
            cb();
        }
    });

    await pipeline(fs.createReadStream(dp), scanner, new PassThrough());
    return offsets;
}

async function generateSafeThumbnail(dp: string, nonce: Buffer, mimeType: string, id: string, key: Buffer | null) {
    const decipher = key ? getDecipherAtOffset(key, nonce, 0) : null;
    const source = fs.createReadStream(dp);
    const stream = new PassThrough();
    if (decipher) pipeline(source, decipher, stream).catch(() => {});
    else pipeline(source, stream).catch(() => {});
    await generateThumbnail(stream as any, id, mimeType);
}

export async function finalizeVaultItem(
    tempDir: string,
    originalName: string,
    nonce: Buffer,
    totalSize: number,
    encLevel: 0 | 1 | 2,
    shouldRandomize: boolean,
    encKey: Buffer | null,
    userKeyId?: string | null
) {
    const enc_name = path.basename(tempDir);
    const mimeType = mime.lookup(originalName) || "application/octet-stream";
    const finalDir = path.join(VAULT_DIR, (
        mimeType.startsWith("image/") ? "images" :
        mimeType.startsWith("video/") ? "videos" :
        mimeType.startsWith("audio/") ? "music" :
        (mimeType.includes("pdf") || mimeType.includes("document") || mimeType.includes("text")) ? "documents" : "files"
    ), enc_name);

    try {
        await fs.ensureDir(path.dirname(finalDir));
        const meta: FileMeta = {
            id: enc_name,
            name: originalName,
            original: originalName,
            size: totalSize,
            nonce: nonce.toString('base64'),
            type: mimeType,
            created_at: Date.now() / 1000,
            duration: 0,
            status: mimeType.startsWith("video/") ? "processing" : "ready",
            encLevel,
            isEncrypted: encLevel > 0,
            userKeyId: userKeyId || undefined
        };
        await saveMeta(tempDir, meta);
        if (await fs.pathExists(finalDir)) await fs.remove(finalDir);
        let moved = false;
        for (let attempt = 0; attempt < 3; attempt++) {
            try {
                await fs.move(tempDir, finalDir);
                moved = true;
                break;
            } catch (err: any) {
                if (attempt < 2 && (err.code === 'EPERM' || err.code === 'EBUSY')) {
                    await new Promise(r => setTimeout(r, 200 * (attempt + 1)));
                } else { throw err; }
            }
        }
        if (!moved) throw new Error(`Could not move ${tempDir} → ${finalDir}`);
        try {
            const dp = vaultDataPath(finalDir);
            await generateSafeThumbnail(dp, nonce, mimeType, enc_name, encKey);
            const updatedMeta = await getMeta(finalDir);
            updatedMeta.thumb = true;
            await saveMeta(finalDir, updatedMeta);
        } catch (thumbErr) {
            console.warn("[Storage] Non-fatal thumbnail error for", enc_name, thumbErr);
        }
    } catch (err) {
        console.error("[Storage] Finalization fatal error, scrubbing temp:", err);
        if (await fs.pathExists(tempDir)) {
            await fs.remove(tempDir).catch(() => {});
        }
        throw err;
    }
}
